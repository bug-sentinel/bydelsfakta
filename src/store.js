import Vue from 'vue';
import Vuex from 'vuex';
import router from './router';

import districts from './config/geoData/districts';
import bydeler from './config/allDistricts';

Vue.use(Vuex);

const state = {
  compareDistricts: false,
  districts: [],
  districtsGeo: districts,
};

const getters = {
  geoDistricts: state => {
    if (!state.compareDistricts && state.districts.length !== 0) {
      return { ...state.districtsGeo[`${bydeler.find(district => district.key === state.districts[0]).uri}`] };
    } else if (state.districts[0] === 'alle') {
      return { ...state.districtsGeo.oslo };
    }

    const features = state.districts.map(id =>
      state.districtsGeo.oslo.features.find(district => district.properties.id === parseInt(id))
    );

    return {
      ...state.districtsGeo.oslo,
      features,
    };
  },
};

const mutations = {
  ADD_DISTRICT(state, payload) {
    state.compareDistricts = true;
    state.districts = payload;
  },
  SELECT_DISTRICT(state, payload) {
    state.compareDistricts = false;
    state.districts = payload;
  },
  CLEAN_STATE(state) {
    state.compareDistricts = false;
    state.districts = [];
    state.districtsGeo = districts;
  },
};

const actions = {
  addDistrict({ commit }, payload) {
    const districts = payload.district.split('-');
    if (districts.length === 1) {
      if (districts[0] === 'alle') {
        commit('ADD_DISTRICT', districts);
      } else {
        const districtValue = bydeler.find(district => district.uri === districts[0]);
        const districtKey = bydeler.find(district => district.key === districts[0]);
        districtValue === undefined
          ? commit('ADD_DISTRICT', [districtKey.key])
          : commit('SELECT_DISTRICT', [districtValue.key]);
      }
    } else {
      commit('ADD_DISTRICT', districts);
    }

    if (payload.pushRoute) {
      router.currentRoute.params.topic === undefined
        ? router.push({ name: 'District', params: { district: districts.join('-') } })
        : router.push({
            name: 'Topic',
            params: { district: districts.join('-'), topic: router.currentRoute.params.topic },
          });
    }
  },
  cleanState({ commit }) {
    commit('CLEAN_STATE');
  },
};

export default new Vuex.Store({
  strict: true,
  state,
  getters,
  mutations,
  actions,
});
