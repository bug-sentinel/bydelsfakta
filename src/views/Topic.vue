<template>
  <div class="main-container">
    <div class="main-container__cards">
      <template v-for="(card, cardIndex) in topics[topic].cards">
        <graph-card
          :key="`card-${cardIndex}`"
          :settings="card"
          v-if="
            card.show === productionMode && card.show !== null
              ? true
              : card.show === true && productionMode === false
              ? true
              : productionMode === null
          "
        />
      </template>
      <div class="related">
        <h2 class="section-heading">{{ $t('topic.seeMore') }}</h2>
        <div class="topics-grid">
          <v-category
            v-for="(item, relatedIndex) in topics[`${topic}`].related"
            :key="`related-${relatedIndex}`"
            :id="topics[`${item}`].value"
            :category="topics[`${item}`].options.kategori"
            :topic="topics[`${item}`].options.tema"
            :bg-image="topics[`${item}`].options.bgImage"
            :txt-color="topics[`${item}`].options.txtColor"
            :district="$route.params.district"
            :disabled="
              topics[`${topic}`].show === productionMode && topics[`${topic}`].show !== null
                ? false
                : topics[`${topic}`].show === true && productionMode === false
                ? false
                : productionMode !== null
            "
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex';
import VCategory from '../components/VCategory.vue';
import GraphCard from '../components/GraphCard.vue';

import { topics, disabledTopics } from '../config/topics';

export default {
  name: 'Topic',
  components: {
    GraphCard,
    VCategory,
  },
  props: {
    district: {
      type: String,
      required: true,
    },
    topic: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      topics: topics,
      disabledTopics: disabledTopics,
    };
  },

  computed: {
    ...mapState(['productionMode']),
    ...mapGetters(['geoDistricts']),
  },
};
</script>

<style scoped>
.related {
  width: 100%;
}
</style>
