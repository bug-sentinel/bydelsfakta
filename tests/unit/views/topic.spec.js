import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import VueResize from 'vue-resize';

import Topic from '../../../src/views/Topic.vue';
import clickOutside from '../../../src/directives/clickOutside';
import store from '../../../src/store';
import setupI18n from '../../../src/i18n'

const i18n = setupI18n();

describe('Topic', () => {
  let wrapper = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    localVue.use(VueResize);
    localVue.use(Vuex);
    localVue.directive('click-outside', clickOutside);
    const router = new VueRouter();
    wrapper = shallowMount(Topic, {
      propsData: {
        district: 'bydel',
        topic: 'alder',
      },
      localVue,
      router,
      store,
      i18n,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders topic-component and finds main-container class', () => {
    expect(wrapper.classes('main-container')).toBe(true);
  });

  test('renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
