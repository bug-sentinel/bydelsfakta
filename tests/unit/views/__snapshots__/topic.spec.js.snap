// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic renders component correctly 1`] = `
<div
  class="main-container"
>
  <div
    class="main-container__cards"
  >
    <graph-card-stub
      settings="[object Object]"
    />
    <graph-card-stub
      settings="[object Object]"
    />
    <graph-card-stub
      settings="[object Object]"
    />
     
    <div
      class="related"
    >
      <h2
        class="section-heading"
      >
        Se også
      </h2>
       
      <div
        class="topics-grid"
      >
        <v-category-stub
          bgcolor="rgb(255, 129, 116)"
          bgimage="/img/folkemengde"
          category="Befolkning"
          disabled="true"
          district="$route.params.district"
          id="folkemengde"
          topic="Folkemengde"
          txtcolor="rgb(245, 173, 165)"
        />
        <v-category-stub
          bgcolor="rgb(66, 248, 182)"
          bgimage="/img/husholdninger"
          category="Bomiljø"
          disabled="true"
          district="$route.params.district"
          id="husholdninger"
          topic="Husholdninger"
          txtcolor="rgb(199, 247, 201)"
        />
        <v-category-stub
          bgcolor="rgb(248, 198, 107)"
          bgimage="/img/levekaar"
          category="Levekår"
          disabled="true"
          district="$route.params.district"
          id="levekaar"
          topic="Levekår"
          txtcolor="rgb(244, 206, 138)"
        />
      </div>
    </div>
  </div>
</div>
`;
